### PR TITLE
[Doc] Add Configuring Batch Predictor section in Predictors User Guides

### PR DESCRIPTION
**Pull request by woshiyyya**
Signed-off-by: Yunxuan Xiao <yunxuanx@Yunxuans-MacBook-Pro.local>

## Why are these changes needed?

- Fixed [this example](https://ray--32436.org.readthedocs.build/en/32436/ray-air/predictors.html#image) by providing the `num_gpus_per_worker` param to ensure the BatchPredictor move data & model to GPU.
- Add a section to highlight the usage of configuring compute resource for BatchPredictor.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by xwjiang2010 — APPROVED** [review]
Thanks!

File: doc/source/ray-air/doc_code/predictors.py [hunk 1]
```diff
@@ -76,6 +76,32 @@ def calculate_accuracy(df):
...
+# Use GPUs for inference: Use 2 workers, each with 1 GPU
+predictions = batch_predictor.predict(
+    ds,
+    feature_columns=["feature_1"],
+    min_scoring_workers=2,
+    max_scoring_workers=2,
+    num_gpus_per_worker=1,
+)
+
+# Auto-scale your workers: each with 3 CPUs and 1 GPU
```
Comment (justinvyu) [inline doc/source/ray-air/doc_code/predictors.py:99]: How is this one "auto-scale" your workers vs. the other ones? Is there no cap on number of scoring workers here?

File: doc/source/ray-air/doc_code/predictors.py [hunk 2]
```diff
@@ -76,6 +76,32 @@ def calculate_accuracy(df):
...
+# Use CPUs for inference: Scaling from 1 to 4 workers, each with 3 CPUs
+predictions = batch_predictor.predict(
+    ds,
+    feature_columns=["feature_1"],
+    min_scoring_workers=1,
+    max_scoring_workers=4,
+    num_cpus_per_worker=3,
+)
+
+# Use GPUs for inference: Use 2 workers, each with 1 GPU
```
Comment (justinvyu) [inline doc/source/ray-air/doc_code/predictors.py:90]:
```suggestion
# Use exactly 2 workers, each performing inference with 1 CPU (default) and 1 GPU
```

File: doc/source/ray-air/doc_code/predictors.py [hunk 3]
```diff
@@ -76,6 +76,32 @@ def calculate_accuracy(df):
 # Final accuracy:  0.5
 # __compute_accuracy_end__
 
+
+# __configure_batch_predictor_start__
+# Use CPUs for inference: Scaling from 1 to 4 workers, each with 3 CPUs
```
Comment (justinvyu) [inline doc/source/ray-air/doc_code/predictors.py:81]:
```suggestion
# Use between 1 to 4 workers, depending on the dataset size
# Each worker performs inference with 3 CPUs (no GPU)
```

**Review by justinvyu — CHANGES_REQUESTED** [review]
Looks good, a few suggestions:

Can we also fix a typo in the bullet point under `separate_gpu_stage`?

```diff
-You may disable it if your preprocessor is very light-weighted.
+You may disable it if your preprocessor is very lightweight.
```

**Comment by amogkam** [comment]
Let’s hold off on merging— I’ll review later today! Thanks @woshiyyya!

File: doc/source/ray-air/doc_code/predictors.py (hunk 1 above)
Comment (woshiyyya) [reply doc/source/ray-air/doc_code/predictors.py:100]: Yes. If no min/max workers provided by user, the BatchPredictor will create an `ActorPoolStrategy` which scaling from 1 to inf.

Comment (amogkam) [reply doc/source/ray-air/doc_code/predictors.py:100]: Agree with @justinvyu here. It's a subtle point, but the other examples are autoscaling as well. Just with different min and max values.

File: doc/source/ray-air/predictors.rst [hunk 4]
```diff
@@ -75,6 +75,41 @@ Additionally, you can compute metrics from the predictions. Do this by:
...
+
+- `separate_gpu_stage`:
+
+  - If using GPUs, whether to use separate stages for GPU inference and data preprocessing.
+
+  - Enabled by default to avoid excessive preprocessing workload on GPU workers. You may disable it if your preprocessor is very light-weighted.
+
+Here are some examples:
+
+.. literalinclude:: doc_code/predictors.py
```
[the remark this answers is not in the document]
Comment (amogkam) [reply doc/source/ray-air/predictors.rst:107]: +1

**Review by amogkam — COMMENTED** [review]
Thanks @woshiyyya! This is very thorough and thanks for going the extra step in updating the documentation as well.

Lgtm once @justinvyu's comments are addressed.

File: doc/source/ray-air/doc_code/predictors.py (hunk 1 above)
Comment (woshiyyya) [reply doc/source/ray-air/doc_code/predictors.py:100]: I see. I'll fix it

File: doc/source/ray-air/predictors.rst [hunk 5]
```diff
@@ -104,10 +104,39 @@ To configure the computation resources for your `BatchPredictor`, you have to se
 
 Here are some examples:
 
+**1. Use CPUs for Batch prediction:**
```
Comment (amogkam) [inline doc/source/ray-air/predictors.rst:107]: Use multiple CPUs for Batch prediction

File: doc/source/ray-air/predictors.rst [hunk 6]
```diff
@@ -104,10 +104,39 @@ To configure the computation resources for your `BatchPredictor`, you have to se
...
+
+- Scale from 1 to 4 workers, depending on the dataset size.
+
+- Each worker has 3 CPUs for inference.
+
+.. literalinclude:: doc_code/predictors.py
+    :language: python
+    :start-after: __configure_batch_predictor_1_start__
+    :end-before: __configure_batch_predictor_1_end__
```
Comment (amogkam) [inline doc/source/ray-air/predictors.rst:116]: Can we keep these sections isolated?

One section only shows using multiple CPUs. One section only shows using GPUs. One section only shows configuring actor scaling.

Comment (woshiyyya) [reply doc/source/ray-air/predictors.rst]: Agree. I've changed that!

**Review by amogkam — APPROVED** [review]

**Comment by woshiyyya** [comment]
> Looks good, a few suggestions:
> 
> Can we also fix a typo in the bullet point under `separate_gpu_stage`?
> 
> ```diff
> -You may disable it if your preprocessor is very light-weighted.
> +You may disable it if your preprocessor is very lightweight.
> ```

Done!

**Review by justinvyu — CHANGES_REQUESTED** [review]
The `torch_image_batch_pretrained` example is timing out on CI. It doesn't seem to have access to enough GPUs.

https://buildkite.com/ray-project/oss-ci-build-pr/builds/12367#018653e5-a9e4-4f2f-8caf-93a0bcfc4238

**Review by justinvyu — APPROVED** [review]
Thanks!

File: doc/BUILD [hunk 7]
```diff
@@ -211,3 +214,16 @@ py_test_run_all_subdirectory(
...
+# Run GPU tests
+# --------------
+
+py_test_run_all_subdirectory(
+    size = "large",
+    include = ["source/ray-air/doc_code/predictors.py"],
+    exclude = [],
+    extra_srcs = [],
+    tags = ["exclusive", "team:ml", "ray_air", "gpu"],
+)
```
Comment (jiaodong) [inline doc/BUILD:229]: newline ?

**Comment by amogkam** [comment]
@woshiyyya, can we get the tests to pass here?

**Review by richardliaw — APPROVED** [review]

**Pull request merged**